File: scripts/check-factory-layout.py
#!/usr/bin/env python3
"""Offline structural regression check for the initialized factory layout."""

import shutil
import tempfile
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
CONTRACT = ROOT / "docs" / "factory-layout.md"

ROOT_FILES = {
    ".gitignore",
    "AGENT.md",
    "CLAUDE.md",
    "README.md",
    "start.py",
    "docs/bindings.md",
}
ROOT_DIRS = {".factory", ".github", ".opencode", ".claude", ".pi", "docs"}
APPLICATION_DIRS = {"app", "cmd", "config", "lib", "migrations", "public", "src", "test", "tests"}
TOOLCHAIN_FILES = {
    "Cargo.lock",
    "Cargo.toml",
    "Gemfile",
    "composer.json",
    "go.mod",
    "go.sum",
    "package-lock.json",
    "package.json",
    "pnpm-lock.yaml",
    "pom.xml",
    "pyproject.toml",
    "requirements.txt",
    "tsconfig.json",
    "yarn.lock",
}
FACTORY_DIRS = {"checks", "docs", "hooks", "scripts", "templates"}
LEGACY_SUPPORT_DIRS = {"checks", "hooks", "scripts", "templates"}
LEGACY_SUPPORT_FILES = {
    "docs/agent-init.md",
    "docs/bootstrap.md",
    "docs/determinism.md",
    "docs/engineering-handbook.md",
    "docs/github-governance.md",
    "docs/org-factory.md",
    "docs/workflow.md",
}
GENERATED_FILES = {
    ".github/workflows/ci.yml",
    ".opencode/plugins/factory-start.ts",
    "docs/bindings.md",
}


def _relative_entries(root):
    return {path.relative_to(root).as_posix() for path in root.rglob("*")}


def check(root):
    """Return structural errors for a fresh initialized-layout fixture."""
    root = Path(root)
    entries = _relative_entries(root)
    errors = []

    for path in sorted(ROOT_FILES):
        if not (root / path).is_file():
            errors.append("required root file missing: %s" % path)
    for path in sorted(GENERATED_FILES):
        if not (root / path).is_file():
            errors.append("generated output missing: %s" % path)

    root_entries = {path.split("/", 1)[0] for path in entries}
    allowed_root_entries = ROOT_FILES | ROOT_DIRS | APPLICATION_DIRS | TOOLCHAIN_FILES
    for entry in sorted(root_entries - allowed_root_entries):
        errors.append("root path is outside the canonical allowlist: %s" % entry)
    for entry in sorted(LEGACY_SUPPORT_DIRS & root_entries):
        errors.append("factory support directory remains at root: %s" % entry)
    for path in sorted(LEGACY_SUPPORT_FILES & entries):
        errors.append("factory support file remains at root: %s" % path)

    factory = root / ".factory"
    if not factory.is_dir():
        errors.append("missing factory boundary: .factory")
    else:
        children = {path.name for path in factory.iterdir()}
        for child in sorted(FACTORY_DIRS - children):
            errors.append("missing .factory directory: .factory/%s" % child)
        for child in sorted(children - FACTORY_DIRS):
            errors.append("unexpected .factory entry: .factory/%s" % child)
        if (factory / "layout.json").exists():
            errors.append(".factory must not contain a second layout manifest")

    text = CONTRACT.read_text(encoding="utf-8")
    for path in sorted(ROOT_FILES | GENERATED_FILES):
        if "`%s`" % path not in text:
            errors.append("contract omits path: %s" % path)
    for path in sorted(".factory/" + child + "/" for child in FACTORY_DIRS):
        if path not in text:
            errors.append("contract omits directory: %s" % path)
    return errors


def _write_fixture(root):
    for path in ROOT_FILES | GENERATED_FILES:
        path = root / path
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text("fixture\n", encoding="utf-8")
    for directory in ROOT_DIRS | APPLICATION_DIRS:
        (root / directory).mkdir(parents=True, exist_ok=True)
    for directory in FACTORY_DIRS:
        path = root / ".factory" / directory
        path.mkdir(parents=True, exist_ok=True)
        (path / "fixture.txt").write_text("fixture\n", encoding="utf-8")
    (root / "src" / "main.py").write_text("# application fixture\n", encoding="utf-8")


def self_check():
    with tempfile.TemporaryDirectory() as directory:
        root = Path(directory)
        _write_fixture(root)
        assert check(root) == []

        (root / "scripts").mkdir()
        errors = check(root)
        assert any("remains at root: scripts" in error for error in errors), errors
        shutil.rmtree(root / "scripts")

        (root / "docs" / "workflow.md").write_text("fixture\n", encoding="utf-8")
        errors = check(root)
        assert any("support file remains at root: docs/workflow.md" in error for error in errors), errors
        (root / "docs" / "workflow.md").unlink()

        shutil.rmtree(root / ".factory" / "templates")
        errors = check(root)
        assert any("missing .factory directory: .factory/templates" in error for error in errors), errors
        (root / ".factory" / "templates").mkdir()

        (root / ".factory" / "layout.json").write_text("{}\n", encoding="utf-8")
        errors = check(root)
        assert any("second layout manifest" in error for error in errors), errors
    print("factory layout structural self-check OK")


if __name__ == "__main__":
    self_check()

File: scripts/check-factory-layout.mjs
#!/usr/bin/env node
import assert from "node:assert/strict";
import { lstat, mkdir, mkdtemp, readFile, readdir, rm, writeFile } from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

const requiredRoot = [".gitignore", "AGENT.md", "CLAUDE.md", "README.md", "start.mjs", "docs/bindings.md"];
const optionalOutputs = [".github/workflows/ci.yml"];
const requiredSupport = ["docs", "hooks", "scripts", "templates"];
const optionalSupport = ["checks"];
const layoutContract = ".factory/docs/factory-layout.md";
const inheritedFiles = [layoutContract, ".factory/hooks/README.md", ".factory/scripts/check-factory-layout.mjs", ".factory/templates/agent-runbook.md"];

const walk = async (directory, relative = "") => {
  const result = [];
  for (const entry of await readdir(directory, { withFileTypes: true })) {
    const name = path.posix.join(relative, entry.name);
    result.push(name);
    if (entry.isDirectory()) result.push(...await walk(path.join(directory, entry.name), name));
  }
  return result;
};

export const check = async (projectRoot) => {
  const entries = new Set(await walk(projectRoot));
  const errors = [];
  const kind = async (relative) => lstat(path.join(projectRoot, relative)).catch((error) => {
    if (error.code === "ENOENT" || error.code === "ENOTDIR") return null;
    throw error;
  });
  for (const file of requiredRoot) {
    if (!(await kind(file))?.isFile()) errors.push(`required root file missing or not a file: ${file}`);
  }
  for (const file of optionalOutputs) {
    if (entries.has(file) && !(await kind(file))?.isFile()) errors.push(`generated output is not a file: ${file}`);
  }
  for (const directory of ["checks", "hooks", "scripts", "templates"]) {
    if (entries.has(directory)) errors.push(`factory support directory remains at root: ${directory}`);
  }
  if (!(await kind(".factory"))?.isDirectory()) {
    errors.push("missing factory boundary: .factory");
  } else {
    for (const directory of [...requiredSupport, ...optionalSupport]) {
      const relative = `.factory/${directory}`;
      const entry = await kind(relative);
      if (requiredSupport.includes(directory) || entry) {
        if (!entry?.isDirectory()) errors.push(`missing or invalid .factory directory: ${relative}`);
      }
    }
    for (const file of inheritedFiles) {
      if (!(await kind(file))?.isFile()) errors.push(`inherited support file missing or not a file: ${file}`);
    }
  }
  const text = await readFile(path.join(projectRoot, layoutContract), "utf8").catch(() => "");
  for (const file of [...requiredRoot, ...optionalOutputs]) {
    if (!text.includes(`\`${file}\``)) errors.push(`contract omits path: ${file}`);
  }
  return errors;
};

const fixture = async (directory) => {
  const paths = [...requiredRoot, ...inheritedFiles];
  for (const file of paths) {
    const target = path.join(directory, file);
    await mkdir(path.dirname(target), { recursive: true });
    await writeFile(target, [...requiredRoot, ...optionalOutputs].map((item) => `\`${item}\``).join("\n"));
  }
  for (const child of requiredSupport) await mkdir(path.join(directory, ".factory", child), { recursive: true });
};

export const selfCheck = async () => {
  const directory = await mkdtemp(path.join(os.tmpdir(), "factory-layout-"));
  const target = (relative) => path.join(directory, relative);
  try {
    await fixture(directory);
    assert.deepEqual(await check(directory), [], "minimal generated layout must pass");

    await mkdir(target(".factory/checks"));
    await mkdir(target(".github/workflows"), { recursive: true });
    await writeFile(target(".github/workflows/ci.yml"), "jobs: {}\n");
    assert.deepEqual(await check(directory), [], "optional checks and CI must pass when present");
    await rm(target(".factory/checks"), { recursive: true });
    await writeFile(target(".factory/checks"), "not a directory\n");
    assert.deepEqual(await check(directory), ["missing or invalid .factory directory: .factory/checks"]);
    await rm(target(".factory/checks"));

    await rm(target(".github/workflows/ci.yml"));
    await mkdir(target(".github/workflows/ci.yml"));
    assert.deepEqual(await check(directory), ["generated output is not a file: .github/workflows/ci.yml"]);
    await rm(target(".github/workflows/ci.yml"), { recursive: true });

    const inheritedChecker = ".factory/scripts/check-factory-layout.mjs";
    await rm(target(inheritedChecker));
    assert.deepEqual(await check(directory), [`inherited support file missing or not a file: ${inheritedChecker}`]);
    await writeFile(target(inheritedChecker), "inherited checker\n");

    await rm(target("docs/bindings.md"));
    assert.deepEqual(await check(directory), ["required root file missing or not a file: docs/bindings.md"]);
    await writeFile(target("docs/bindings.md"), "bindings\n");

    await rm(target(".factory/hooks"), { recursive: true });
    assert.deepEqual(await check(directory), [
      "missing or invalid .factory directory: .factory/hooks",
      "inherited support file missing or not a file: .factory/hooks/README.md",
    ]);
    await mkdir(target(".factory/hooks"));
    await writeFile(target(".factory/hooks/README.md"), "inherited hooks\n");

    await mkdir(target("scripts"));
    assert.deepEqual(await check(directory), ["factory support directory remains at root: scripts"]);
    await rm(target("scripts"), { recursive: true });
    assert.deepEqual(await check(directory), []);
    console.log("factory layout structural self-check OK");
  } finally {
    await rm(directory, { recursive: true, force: true });
  }
};

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) selfCheck().catch((error) => { console.error(error); process.exitCode = 1; });
